### PR TITLE
Make features independent

### PR DESCRIPTION
**Pull request by tlehman**
Note: for existing features, the original behavior is preserved, the existing features (ones currently stored in redis that get returned from `Feature.find`, they have ordered threshold modulus values, like this:

screen shot 2015-08-18 at 12 00 24 pm

# Statistical independence of features
Given a feature X with a users_rate of 60 and a feature Y with a users_rate rate of 10, and a user U, the probability of U seeing feature X is 60%, and 10% for Y.

In notation: 

     P(X) = 0.6
     P(Y) = 0.1

The property we are testing below is independence, which says that the probability of event X has no influence on the probability of event Y. In our example, the probability of seeing both X and Y is 60% * 10%, which is 6%

In notation:

             P(X and Y)     = P(X)P(Y)         = 0.6*0.1 = 0.06
             P(X and not Y) = P(X)(1 - P(Y))   = 0.6*0.9 = 0.54
             P(not X and Y) = (1 - P(X))P(Y)   = 0.4*0.1 = 0.04
             P(not X nor Y) = (1-P(X))(1-P(Y)) = 0.4*0.9 = 0.36
                                                         + -----
             All possibilities should sum to 1.00 ----->   1.00

# UI to check if a user (or visitor) can see a feature:

blackbeard

### CODE REVIEW

File: lib/blackbeard/feature.rb [hunk 1]
```diff
@@ -4,12 +4,22 @@ class Feature < Storable
     include Chartable
 
     set_master_key :features
-    string_attributes :name, :description, :status
+    string_attributes :name, :description, :status, :threshold_array_serialized
```
Comment (rgraff) [inline lib/blackbeard/feature.rb]: I like this approach, but I would make it a son attribute. Also could use a better name. :threshold_moduli ?

File: lib/blackbeard/feature.rb [hunk 2]
```diff
@@ -4,12 +4,22 @@ class Feature < Storable
...
     integer_attributes :visitors_rate, :users_rate
     json_attributes :group_segments
 
     has_many :metrics => 'Metric'
 
+    def threshold_array=(thresholds)
+      self.threshold_array_serialized = thresholds.join(",")
+    end
+
+    def threshold_array
```
Comment (rgraff) [inline lib/blackbeard/feature.rb]: The getter should call the setter so that the first time, it is called the array is generated and saved.

**Comment by rgraff** [comment]
This is a very useful feature that I've been meaning to implement. I like the approach (with comments). Maybe make the threshold moduli visible in the UI as well.  Sometimes, it's useful to determine if a user with who experienced a problem was in the test group.

I need to do a more thorough review still, but I think this is close.

File: lib/blackbeard/feature.rb (hunk 1 above)
Comment (tlehman) [inline lib/blackbeard/feature.rb]: I was wondering about json attributes, I'll try that out.

**Comment by tlehman** [comment]
Cool, I'll address the comments and add some more tests. I need to instantiate a bunch of features and then check to make sure the statistics suggest it's working right.

I'll add the preliminary UI, let me know if you have any mockups or ideas about how to display the information, I have some myself, but I am not experienced in design.

**Comment by tlehman** [comment]
All the statistics look great, just need to build out the UI part.

**Comment by tlehman** [comment]
@rgraff @robotloveskitten This is ready for review when you have a chance.

File: lib/blackbeard/version.rb [hunk 3]
```diff
@@ -1,3 +1,3 @@
 module Blackbeard
-  VERSION = "0.0.5.3"
+  VERSION = "0.0.6.0"
```
Comment (rgraff) [inline lib/blackbeard/version.rb]: leave version as is. I will bump when I release gem.  In meantime, reference by commit in Gemfile.

**Comment by rgraff** [comment]
Don't require `Blackbeard::Feature.all { |feature| feature.save }` Existing features will an no moduli array should create one `(0..99).to_a` that mimics existing behavior.

Update readme

**Comment by tlehman** [comment]
@rgraff @robotloveskitten I have tried this out on stg2, it now preserves the existing behavior, and only uses the randomized arrays on newly created features. So going forward, we will start to see independent features, but we won't abruptly rebucket any users when we deploy this.

**Comment by robotloveskitten** [comment]
sounds cool to me.

File: spec/feature_rollout_spec.rb [hunk 4]
```diff
@@ -1,38 +1,56 @@
...
+        expect(feature.rollout?(context)).to eq(true)
       end
     end
 
-    describe "active_user?" do
+
+    describe "#active_user?" do
       context "with no logged in user" do
-        it "should be true if users_rate is 100" do
+        it "should be false if users_rate is 100" do
```
Comment (kroehre) [inline spec/feature_rollout_spec.rb]: could just do "should be false" here

File: spec/feature_spec.rb [hunk 5]
```diff
@@ -19,7 +80,7 @@ module Blackbeard
     end
 
     describe "#active_for?" do
-      let(:context){ double :unique_identifier => 'bob' }
+      let(:context) { double :unique_identifier => 'bob' }
```
Comment (kroehre) [inline spec/feature_spec.rb]: `double(unique_identifier: 'bob')`

**Comment by kroehre** [comment]
+2

**Pull request closed without merging**